src/coreclr/jit/targetamd64.h

```diff
@@ -207,10 +207,10 @@
   #define RBM_CALLEE_GCTRASH_WRITEBARRIER       RBM_CALLEE_TRASH_NOGC
 
   // Registers killed by CORINFO_HELP_ASSIGN_BYREF.
-  #define RBM_CALLEE_TRASH_WRITEBARRIER_BYREF   (RBM_RSI | RBM_RDI | RBM_CALLEE_TRASH_NOGC)
+  #define RBM_CALLEE_TRASH_WRITEBARRIER_BYREF   (RBM_RSI | RBM_RDI | RBM_RAX | RBM_RCX | RBM_R10 | RBM_R11)
```

Comment:
There is no reason for the write barrier to trash different registers on regular CoreCLR vs. native AOT. It is easy to make it the same.

> Windows-CoreCLR-x64 does touches R10/R11 too, but it push/pop them,

It is only on debug-only GCStress path.

Reply:
Yep, but since 3 out of 4 x64 platforms need them - I decided to include it conservatively into Windows as well. So:

1. `Windows-x64-CoreCLR`:
2. `Windows-x64-Naot`: needs R10, R11
3. `Unix-x64-CoreCLR`: needs R10, R11
4. `Unix-x64-Naot`: needs R10, R11

Added a comment. This costs few missing diff improvements on `Windows-x64-CoreCLR` so I decided to leave as is, the proper fix won't be simple as `#ifdef` since IsNaot + IsWindows/IsUnix checks aren't compile-time.

Reply:
This is not someplace we should be conservative. We should be precise. If Windows-x64-CoreCLR doesn't need R10/R11 in Release, then it shouldn't use them (or at least shouldn't trash them), and the JIT should be free to use them.

You can change the `#define` to call a function that checks the OS/ABI.

Is there any reason why the debug-only code path on Linux can't push/pop R10/R11 as well? (Maybe stack walking prevents it?)

Reply:
There is no reason for the reg alloc requirements for this helper to differ between Windows/Unix or JIT/AOT. The assembly code for the helper can be exactly the same for all variants. We ended up with different tweaks applied to different copies of the assembly code over time and that's why the register requirements differ.

Reply:
> There is no reason for the reg alloc requirements for this helper to differ between Windows/Unix or JIT/AOT. The assembly code for the helper can be exactly the same for all variants. We ended up with different tweaks applied to different copies of the assembly code over time and that's why the register requirements differ.

@jkotas from a quick look, one issue that on windows (.asm) we can do:
```asm
cmp     rdi, [g_lowest_address]
```
while on Linux (.s) we have to do:
```asm
PREPARE_EXTERNAL_VAR g_lowest_address, rax
cmp     rdi, [rax]
```

Reply:
We should be able to the same on Linux too. We do that for native AOT, but not for regular CoreCLR for some reason: https://github.com/dotnet/runtime/blob/8330db998659c4e6410aba370b37e4304a517a2b/src/coreclr/nativeaot/Runtime/amd64/WriteBarriers.S#L181

Reply:
Copy-pasted the CoreCLR-x64-windows version for Linux. Let's see..

Reply:
ah, I should've tested it locally first, forgot about comments // and the syntax for labels